File: service/scaladsl/client/src/main/scala/com/lightbend/lagom/scaladsl/client/CircuitBreakingServiceLocator.scala
/*
 * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
 */
package com.lightbend.lagom.scaladsl.client

import java.net.URI

import com.lightbend.lagom.internal.client.CircuitBreakers
import com.lightbend.lagom.scaladsl.api.Descriptor.Call
import com.lightbend.lagom.scaladsl.api.{ CircuitBreaker, Descriptor, ServiceLocator }

import scala.concurrent.{ ExecutionContext, Future }

/**
 * Abstract service locator that provides circuit breaking.
 *
 * Generally, only the [[ServiceLocator.locate()]] method needs to be implemented, however
 * [[doWithServiceImpl()]] can be overridden if the service locator wants to
 * handle failures in some way.
 */
abstract class CircuitBreakingServiceLocator(circuitBreakers: CircuitBreakers)(implicit ec: ExecutionContext) extends ServiceLocator {

  /**
   * Do the given block with the given service looked up.
   *
   * This is invoked by [[doWithService()]], after wrapping the passed in block
   * in a circuit breaker if configured to do so.
   *
   * The default implementation just delegates to the [[locate()]] method, but this method
   * can be overridden if the service locator wants to inject other behaviour after the service call is complete.
   *
   * @param name        The service name.
   * @param serviceCall The service call that needs the service lookup.
   * @param block       A block of code that will use the looked up service, typically, to make a call on that service.
   * @return A future of the result of the block, if the service lookup was successful.
   */
  protected def doWithServiceImpl[T](name: String, serviceCall: Descriptor.Call[_, _])(block: URI => Future[T]): Future[Option[T]] = {
    locate(name, serviceCall).flatMap {
      case (Some(uri)) => block(uri).map(Some.apply)
      case None        => Future.successful(None)
    }
  }

  override final def doWithService[T](name: String, serviceCall: Call[_, _])(block: (URI) => Future[T])(implicit ec: ExecutionContext): Future[Option[T]] = {
    serviceCall.circuitBreaker.filter(_ != CircuitBreaker.None).map { cb =>
      val circuitBreakerId = cb match {
        case cbid: CircuitBreaker.CircuitBreakerId => cbid.id
        case _                                     => name
      }

      doWithServiceImpl(name, serviceCall) { uri =>
        circuitBreakers.withCircuitBreaker(circuitBreakerId)(block(uri))
      }
    }.getOrElse {
      doWithServiceImpl(name, serviceCall)(block)
    }
  }
}

File: service/javadsl/client/src/main/java/com/lightbend/lagom/javadsl/client/CircuitBreakingServiceLocator.java
/*
 * Copyright (C) 2016-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package com.lightbend.lagom.javadsl.client;

import com.lightbend.lagom.internal.client.CircuitBreakers;
import com.lightbend.lagom.internal.javadsl.client.CircuitBreakersPanelImpl;
import com.lightbend.lagom.javadsl.api.CircuitBreaker;
import com.lightbend.lagom.javadsl.api.Descriptor;
import com.lightbend.lagom.javadsl.api.ServiceLocator;

import java.net.URI;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;

/**
 * Abstract service locator that provides circuit breaking.
 *
 * Generally, only the {@link #locate(String, Descriptor.Call)} method needs to be implemented, however
 * {@link #doWithServiceImpl(String, Descriptor.Call, Function)} can be overridden if the service locator wants to
 * handle failures in some way.
 */
public abstract class CircuitBreakingServiceLocator implements ServiceLocator {

    private final CircuitBreakersPanel circuitBreakersPanel;

    public CircuitBreakingServiceLocator(CircuitBreakersPanel circuitBreakersPanel) {
        this.circuitBreakersPanel = circuitBreakersPanel;
    }

    /**
     * @deprecated Use constructor accepting {@link CircuitBreakersPanel} instead
     * @param circuitBreakers
     */
    @Deprecated
    public CircuitBreakingServiceLocator(CircuitBreakers circuitBreakers) {
        // convert passed Scala CircuitBreaker to a Java one
        this.circuitBreakersPanel = new CircuitBreakersPanelImpl(circuitBreakers);
    }

    /**
     * Do the given block with the given service looked up.
     *
     * This is invoked by {@link #doWithService(String, Descriptor.Call, Function)}, after wrapping the passed in block
     * in a circuit breaker if configured to do so.
     *
     * The default implementation just delegates to the {@link #locate(String, Descriptor.Call)} method, but this method
     * can be overridden if the service locator wants to inject other behaviour after the service call is complete.
     *
     * @param name The service name.
     * @param serviceCall The service call that needs the service lookup.
     * @param block A block of code that will use the looked up service, typically, to make a call on that service.
     * @return A future of the result of the block, if the service lookup was successful.
     */
    protected <T> CompletionStage<Optional<T>> doWithServiceImpl(String name, Descriptor.Call<?, ?> serviceCall, Function<URI, CompletionStage<T>> block) {
        return locate(name, serviceCall).thenCompose(uri -> {
          return uri
                  .map(u -> block.apply(u).thenApply(Optional::of))
                  .orElseGet(() -> CompletableFuture.completedFuture(Optional.empty()));
        });
    }

    @Override
    public final <T> CompletionStage<Optional<T>> doWithService(String serviceName, Descriptor.Call<?, ?> serviceCall, Function<URI, CompletionStage<T>> block) {

        return serviceCall.circuitBreaker()
                .filter(cb -> !cb.equals(CircuitBreaker.none()))
                .map(cb -> {

                    String circuitBreakerId;

                    if (cb instanceof CircuitBreaker.CircuitBreakerId) {
                        circuitBreakerId = ((CircuitBreaker.CircuitBreakerId) cb).id();
                    } else {
                        circuitBreakerId = serviceName;
                    }

                    return doWithServiceImpl(
                            serviceName,
                            serviceCall,
                            uri -> circuitBreakersPanel.withCircuitBreaker(circuitBreakerId, () -> block.apply(uri))
                          );

                })
                .orElseGet(
                  () -> doWithServiceImpl(serviceName, serviceCall, block)
                );
    }
}
